### PR TITLE
Detect if AppX is background

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Related to #9970

While we haven't public API to detect whether AppX is GUI or CUI before the process is started we can make this after the process is started and decide whether to wait for the process or not.

Based on @jborean93's idea - thanks!

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (jborean93) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:671]: I don't think this is a store application specific point but any process that spawns another.  Any process could do this and `wt.exe` is just one of them while `skype.exe` stays alive as that's what they are programmed to do. I think this comment block should state why we get the exe this way as it's to let Windows resolve AppExecLinks rather than use undocumented structure fields.

**Review by jborean93 — COMMENTED** [review]
Nice, I didn't think of just checking the `MainModule` in the `Process` object but that looks like it will fit the bill. While not suspending the process on startup to ensure we get the module info is an annoying limitation I think what you've done is a decent enough workaround to handle cases where the the process had exited soon after it had started.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:689]: Can't you just use the `IsWindowsApplication()` method?

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:689]: We need the workaround just because `IsWindowsApplication()` today doen't work for AppX reparse points.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:671]: I think it makes sense to add info about AppX reparse point API gaps and maybe about our tracking issue. 
Usually @daxian-dbw comes with final great comment :-)

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:671]: Yea I think it's fine to mention this is for `AppExecLink` workaround but I think the extra info around wt and skype is just not needed, especially since the behaviour mentioned isn't actually related to AppExecLinks themselves. Usually just keep it concise is worthwhile IMO. But I'll leave it to @daxian-dbw to have the final say.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:671]: > the behaviour mentioned isn't actually related to AppExecLinks

The whole workaround is needed only for Appx - other scenarios are covered by `IsWindowsApplication()` early.

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:671]: I understand, I just mean the whole `wt.exe` then calls `WindowsTerminal.exe` then exits and `skype.exe` just continues to run isn't something specific to AppExecLinks but just how those executables work. The AppExecLink behaviour not covered by `IsWindowsApplication()` is just how Windows interprets the reparse point itself and automatically repoints the final module to it's real target. I'm suggesting that the comment mentions that part rather than what's there currently.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:689]: I'm not saying relying on `isWindowsApplication` defined way above, but using `mainModuleFileName` passed to `IsWindowsApplication()`.  Isn't lines 689-690 doing essentially the same thing?

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:689]: Ah, yes, we can:-)

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: Done.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:678]: I updated the comment a bit.

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:678]: Looks great.

**Review by SteveL-MSFT — APPROVED** [review]
Great way to check without relying on undocumented API

**Comment by iSazonov** [comment]
> Great way to check without relying on undocumented API

It's @jborean93 idea. Thanks!

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:676]: Thanks for putting up the fix!
I left some comments above, but my biggest question mark is whether it's worth doing this given that we will change to the public API once it's available in the Windows App SDK, or we are not going to use that public API anymore?

With the current fix, it's still incomplete in that we will allocate console when it's unnecessary (check out the use of `IsWindowsApplication`), and this has to be done before the exe starts. This doesn't sound like a big deal because this likely matters only when a non-console application is hosting PowerShell, but it's surely incomplete.

The upside to stick with this fix is that we can avoid the dependency on the Windows App SDK, and that's actually not bad from the maintenance point of view ...

However, if this is meant to be a temp workaround and we are going to depend on Windows App SDK, then a new issue needs to be opened to track that because the current tracking issue #9970 will be closed by this fix.

@iSazonov @jborean93 @SteveL-MSFT what do you think?

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:676]: > With the current fix, it's still incomplete in that we will allocate console when it's unnecessary

What's the purpose of this, if Windows was to start a console application and one wasn't already allocated it will do so automatically. A brief look at the code shows it might be to hide the Window but surely the .NET Process [has this exposed already](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.windowstyle?view=net-5.0#System_Diagnostics_ProcessStartInfo_WindowStyle). I also don't think we need to care about redirection with stdio as that is governed by whether there is input or the output is redirected somewhere which is down to the code rather than the executable being used.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:669]: It is Unix code but ok, will change :-)

Done

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:688]: Done.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:701]: What is changed here?

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:703]: Debugger shows me `Win32Exception` if the process was exited.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > > With the current fix, it's still incomplete in that we will allocate console when it's unnecessary
> 
> What's the purpose of this, if Windows was to start a console application and one wasn't already allocated it will do so automatically. A brief look at the code shows it might be to hide the Window but surely the .NET Process [has this exposed already](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.windowstyle?view=net-5.0#System_Diagnostics_ProcessStartInfo_WindowStyle). I also don't think we need to care about redirection with stdio as that is governed by whether there is input or the output is redirected somewhere which is down to the code rather than the executable being used.

If an user run a console application the user expects to get output from the application. This is why pwsh always creates a console (whether it is a console application or not) before run a console application.
The fact force me think we should _always_ allocate a console and grab its output.
I believe we could do this for GUI apps too - if GUI allocates a console _why do it so_? (Skype (from Store) writes something to a console at start, I don't know whether it allocate a console always.) At least we could do that until we have a way to always identify the type of application before running. Nevertheless if GUI app _inherits_  a console and writes something to the console the design assumes they want we read the output; otherwise the GUI application shouldn't inherits a console but allocate new one if needed.

Also we should add redirection [here](https://github.com/PowerShell/PowerShell/blob/f2a42fb9644303ab382297b10efc5acc708966b7/src/System.Management.Automation/engine/NativeCommandProcessor.cs#L596-L600) too for consistency.
Also we should/could do the same for associated file scenario: if we create `file.findstr` file, create assoc for *.findstr with `findstr /f:` and run `file.findstr` we'd want to grab the output.

> I left some comments above, but my biggest question mark is whether it's worth doing this given that we will change to the public API once it's available in the Windows App SDK, or are we not going to use that public API anymore?

The best fix would be in SHGetFileInfo so that it works with AppX. Otherwise we have to do one p/invoke more.
But I guess this would be never backported to early Windows 10 versions.
And if the fix will be created it will be only in Windows 11. It is not helps us since we should support old Windows versions.
The same for Windows App SDK - will we get the API for Windows 11 only? I think yes...

So we have two questions:
1. Console allocations and output grabbing
As I said we could always allocate a console and always grab app output since if app does output it assumes the output is by-design and we should read it.
2. Waiting application exit.
The fix resolves the issue but better fix would be in SHGetFileInfo for all Windows 10.
One thing more we could add - it is a cache in IsWindowsApplication() - no need to check the same application every time we run it (in a cycle).

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > If an user run a console application the user expects to get output from the application. This is why pwsh always creates a console (whether it is a console application or not) before run a console application.

So there are 4 scenarios here that I see:

* console powershell spawning console app
  * There will be a conhost associated with powershell and the console app inherits this
* console powershell spawning GUI app
  * There will be a conhost associated with powershell
  * I can't 100% remember if the GUI app still inherits this but it doesn't really matter here
* GUI powershell spawning console app
  * There is no conhost associated with powershell
  * When powershell starts the console app Windows will automatically assign a new conhost for it
* GUI powershell spawning console app
  * There is no conshost associated with powershell and the GUI app doesn't need one

The important one here is the 3rd scenario as currently PowerShell will allocate a new console to itself and have the subprocess inherit it. What confuses me is why do this at all when Windows will automatically create the conhost for a console app if there isn't one to inherit. This is how you can start a console process from a GUI app, say `explorer.exe`, and a new console appears.

When you start talking about redirection and having PowerShell capture (or send input) then this isn't related to the console itself. When a process is created there is a structure called [STARTUPINFO](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfow) which defines a few things about the new process but in particular there is the `hStdInput`, `hStdOutput`, `hStdError` fields. These fields are used by the parent process to override the `stdin`, `stdout`, `stderr` respectively of the new process allowing it to write/read data on the pipes specified. You can define this for any type of application and regardless of whether the child process is a GUI, console app, creates its own console, or inherits it, the stdio handles of that child will reflect what's defined here.

This is what allows PowerShell to capture the output of a process and is totally independent on the console that the child host is using. The only caveat here is if the child process is writing to the console directly using something like [WriteConsoleOutput](https://docs.microsoft.com/en-us/windows/console/writeconsoleoutput). This doesn't use the stdio pipes defined on the process but rather writes directly to the console buffer skipping the redirection. If the child process inherits the console from powershell it will appear in that console buffer, otherwise it will appear in the conhost that was spawned with the child. PowerShell can't capture/redirect this type of output right now, technically it could use the `ConPTY` APIs that Windows Terminal uses but that's orthogonal to this particular problem so let's ignore that.

The only purpose I can see with the code to allocate a console is to hide it from the user but that's already achievable with the [ProcessStartInfo.WindowStyle](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.windowstyle?view=net-5.0#System_Diagnostics_ProcessStartInfo_WindowStyle) that I linked to before. It could technically be used to ensure the child process using an encoding known to PowerShell but this isn't done today and opens a whole can of worms.

In the end I might be missing something here but when looking at the code it sounds like we are creating the console for a scenario that Windows handles for us anyway. I personally believe this whole thing could be removed and @daxian-dbw point becomes moot based on my understanding but happy to be proven wrong.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > When you start talking about redirection and having PowerShell capture (or send input) then this isn't related to the console itself. 

It is true in common but we should consider how PowerShell code works and it mixes console and redirection :-) See `AlwaysCaptureApplicationIO`. It would be a braking change for hosting apps if we removed this.

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: PowerShell can't capture direct console output unless it starts using the ConPTY APIs introduced in Windows 10. AFAIK right now it only captures output using pipe redirection which isn't related to a console being present or not.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: I say about how PowerShell code works - if it creates a console it turns on redirection too.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: I wonder the logic because we allocate the hidden console once and don't remove it so on next iteration the console will be detected and no redirection will be turned on.

Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: Based on discussions with AppX team it's unlikely we'll have public api or documentation of their struct

**Comment by iSazonov** [comment]
@daxian-dbw Is anything I should fix more?

**Comment by Hffggvb** [comment]
Lev me alone

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: Sorry for the delay of my response. The code that creates a new console has been there for a very long time. By looking at the code, it seems one thing it does there is to keep the focus on the current active application (`GetForegroundWindow` before creating the console, and `SetForegroundWindow` afterward). Maybe that's why it creates the console instead of just leave the work to OS.

As for `AlwaysCaptureApplicationIO`, I have no idea why it's needed ...

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:684]: When `_nativeProcess.HasExited == true`, `_isRunningInBackground` is `true` even if it's really a console application.
However, when `_isRunningInBackground == true`, the [code](https://github.com/PowerShell/PowerShell/blob/f922883757dd2d461097b72c0d6aba7584302cba/src/System.Management.Automation/engine/NativeCommandProcessor.cs#L813-L889) that set `LastExitCode` won't run. That code also includes some special handling to transcribing a native application.

It looks to me we still need to know whether an application is console app before starting it, in order to cover the cases where the child process has exit when code reaches here.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: @SteveL-MSFT, the AppX team agreed to provide a new API for querying whether an appx is console app or windows app, and the new API will be shipped in [Windows App SDK](https://github.com/microsoft/WindowsAppSDK).

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > Maybe that's why it creates the console instead of just leave the work to OS.

There is the comment:

                 // On Windows desktops, if the command to run is a console application,
                // then allocate a console if there isn't one attached already...

I understand this as it is specific _host application_ - it is in a console app format but no console is allocated. 
This means that this host does not want the console application being launched to be visible to the user. If we do not create a hidden console, the OS will automatically create a console that will be visible.

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > This means that this host does not want the console application being launched to be visible to the user. If we do not create a hidden console, the OS will automatically create a console that will be visible.

This should be controlled by [ProcessStartInfo.WindowStyle](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.windowstyle?view=net-6.0). By specifying a hidden window then the new conhost will exist but not appear to the end user.

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: Just to prove this point run the following in ISE (or any other console less app that hosts PowerShell)

```powershell
$psi = [System.Diagnostics.ProcessStartInfo]@{
    FileName = 'powershell.exe'
    WindowStyle = 'Hidden'
}
[System.Diagnostics.Process]::Start($psi)
```

No new window will appear but you can see that the process was spawned with it's own console.

image

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: Rather run ping.exe in ISE :-)

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: @jborean93 With that example, the focus is still on ISE after running `[System.Diagnostics.Process]::Start($psi)`, so it looks like the focus will not change when the console window is hidden. Then that code does look unnecessary. The mysterious part is `AlwaysCaptureApplicationIO`, I just cannot make sense from it 😕

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:684]: @iSazonov You may have missed this comment.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > Rather run ping.exe in ISE :-)

I think that's the current code up for discussion, PowerShell will spawn the console and thus conhost leaving it run in the background rather than having it tied to the new process in general. At least when using your ping example through my snippet it acts exactly the same as the `powershell.exe` one. If you were to run it directly like `ping.exe server` then you can see the conhost survives the process due to PowerShell being the one that creates it. IMO this just adds more evidence that PowerShell creating the console if there isn't one isn't the right thing to do and thus is whole point moot.

> The mysterious part is AlwaysCaptureApplicationIO, I just cannot make sense from it confused

Yea I cannot explain why it would be necessary, it seems to be set to true whenever the console PowerShell spawns is hidden but honestly that doesn't sound like a good reason. The only time I think PowerShell should be explicitly capturing data is:

* The output is being redirected for x reason
  * Stored in a var
  * Piped to another command
  * Redirected to another stream/var/null
* In a PSRemoting/server side session where the native app cannot be allowed to interact with the console directly
  * All output must come through the PS streams
  * This is required so that 1 the output isn't lost and 2 the output doesn't break the PSRemoting communication

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > hidden console ... This should be controlled by `ProcessStartInfo.WindowStyle`.

@jborean93 I understand this can be handled by `ProcessStartInfo.WindowStyle`, but how would PowerShell know when it should enforce `WindowStyle.Hidden`? It would still have to do a `GetConsoleWindow` call to see if there is a console already allocated, right?

As for `AlwaysCaptureApplicationIO`, I looked a little more, and it seems to be used to cache the state that "PowerShell is running in a GUI application", and PowerShell should always redirect the stdout and stderr for a console application in such a case, because the new console is hidden, and the output of the application won't be visible if it's not redirected.

Comment (jborean93) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > how would PowerShell know when it should enforce WindowStyle.Hidden? It would still have to do a GetConsoleWindow call to see if there is a console already allocated, right?

Ahh I think it finally clicked for me, yes you would only want to hide new console application windows and not GUI ones and thus need to know beforehand whether it should be hidden or not. Technically you could always hide it and just show the window after it's started once you know it's a GUI and not a console app but I'm unsure whether this is practical or whether there are other problems with this approach.

> I looked a little more, and it seems to be used to cache the state that "PowerShell is running in a GUI application",  and PowerShell should always redirect the stdout and stderr for a console application in such a case

I guess that makes some sense but I'm still unsure whether it's needed for PowerShell to explicitly create the console. If PowerShell knows it doesn't have a console and thus should always redirect and capture the output it doesn't actually need to spawn the console itself. It can just add the stdio redirection when starting the process and still check after it has started whether it needs to block the pipe and output the data like `my.exe | Out-String`.

Ultimately if what you said is that there will be a public API at some point then keeping the current behaviour and using that sounds like a better idea. I still think it would be a good idea to use the undocumented struct for older Windows versions (where you know it hasn't changed) and the new API on ones where it is available.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:672]: > I'm still unsure whether it's needed for PowerShell to explicitly create the console. If PowerShell knows it doesn't have a console and thus should always redirect and capture the output it doesn't actually need to spawn the console itself.

Yes, it may not be needed for PowerShell to explicitly create the console, but the dev who wrote this code just decided to do it so setting `SW_HIDE` and foreground window can all be done at the same place. And maybe it has a slightly perf win given that this console will be present throughout the lifecycle of the process that hosts PowerShell so that the OS doesn't need to allocate/destroy a console for every console app that would be started by PowerShell? Don't know, but I'd rather to keep the code as is since we are able to make sense from it now 😄

> Ultimately if what you said is that there will be a public API at some point then keeping the current behaviour and using that sounds like a better idea. I still think it would be a good idea to use the undocumented struct for older Windows versions (where you know it hasn't changed) and the new API on ones where it is available.

Yeah, I'm also leaning to waiting on the new public API, but don't have a strong opinion. We cannot use the undocumented struct in our code base because of compliance.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:684]: This is the zugzwang :-) Perhaps (I hope) we have made best what we can for starting native application but you ask about termination. The best we can do for this is to think about specific scenarios. 
Difficults are (1) we can catch a race condition with process terminating, (2) even if we could always tell if an application is a console application, this does not guarantee the expected behavior (since any application can create/attach/detach/delete a console by itself).

So it makes sense only think about typical scenarios. Do you know such scenarios we should consider?

As side notice, we could use CREATE_SUSPENDED flag in the dwCreationFlags parameters of the CreateProcess() on Windows to address a race condition with process start..

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:684]: The scenario is where a native console application exits very early. I don't have a working example to demonstrate that, but it surely could happen. I cannot approve this PR with this breaking change.

Using `CreateProcess()` means a fundamental refactoring of the native command processor, which I don't think will be approved. At this point, I think it's better we just wait for the new public API from the AppX team.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: I removed `_nativeProcess.HasExited` from the condition - it is not critical here. Also I rebased to get latest changes.

As for refactoring, I feel this is inevitable. The new Appx API won't solve all the problems. It might make sense for PowerShell team to work with cmd.exe team to find a better way to run apps. As result, Process class could return the information whether running app is GUI or CUI.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: @daxian-dbw Friendly ping...

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]:
```
                        catch (Win32Exception)
                        {
                            // We can not get MainModule, assume that the process has already been completed.
                            // The best thing we can do is to use default behavior like Windows PowerShell.
                            _isRunningInBackground = true;
                        }
```
This is the same effect as having `_isRunningInBackground = ... || _nativeProcess.HasExited`, isn't it? And thus, the issue I brought up above still exists.

I think we should just wait on the new API for detecting whether AppX app is CUI or GUI, instead of dealing with the subtle issues that could be caused by this change.

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: > This is the same effect as having _isRunningInBackground = ... || _nativeProcess.HasExited, isn't it? And thus, the issue I brought up above still exists.

No, this is previous code (Windows PowerShell) default.

I have no strict desire to merge the PR so you can close if you want.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: The `_isRunningInBackground = true;` is previous existing code, but it was not in a `catch (Win32Exception)` block previously . The previous related code was:
```c#
      _isRunningInBackground = true;
      if (!startInfo.UseShellExecute)
      {
          _isRunningInBackground = isWindowsApplication;
      }
```
I will close this PR then, but thank you for trying out this approach. I appreciate your time!

Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: > I appreciate your time!

This is my hobby :-) I'm more concerned about wasting your time, sorry.

**Pull request closed without merging**
